### ISSUE TITLE
t.n.hosts.searchFileFor does not close the file

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4540]
|[<img alt="fijal's avatar" src="https://avatars.githubusercontent.com/u/85944?s=50" width="50" height="50">](https://github.com/fijal)| @fijal reported|
|-|-|
|Trac ID|trac#4540|
|Type|defect|
|Created|2010-07-02 10:34:47Z|
|Branch|https://github.com/twisted/twisted/tree/close-hosts-4540|

Which leads to -1 EMFILE (Too many open files) on non-refcounting based python implementations (like PyPy).

Attachments:

* [t.n.hosts.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/21c/21cb91fe07368a0f6619722b30f4f11916fa53b5/17ba27f84ac191b519feb66ef5409bf24b628ba1.diff) (1131 bytes) - added by fijal on 2010-07-02 10:36:13Z - Patch
* [out.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/21c/21cb91fe07368a0f6619722b30f4f11916fa53b5/3241506a07ba74d27a5dffd4bcecd2dfeeb18577.diff) (1996 bytes) - added by fijal on 2010-07-04 20:21:42Z - 
* [out.2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/21c/21cb91fe07368a0f6619722b30f4f11916fa53b5/eff931f171e6d1759a1b39060fdc3dac77a0dbf4.diff) (2195 bytes) - added by fijal on 2010-07-05 14:34:10Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4540 4540
type__defect defect
reporter__fijal fijal
priority__normal normal
milestone__ 
branch__branches_close_hosts_4540 branches/close-hosts-4540
branch_author__fijal fijal
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1278066887000000 1278066887000000
changetime__1278342833000000 1278342833000000
version__None None
owner__ 
cc__fijal
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="fijal's avatar" src="https://avatars.githubusercontent.com/u/85944?s=50" width="50" height="50">](https://github.com/fijal)|@fijal removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_3"></a>|@mwhudson commented|
|-|-|

Boringly, ReviewProcess requires that this patch adds tests and API docs for searchFileFor.  Neither should be very hard though.

It also needs a file describing the change in twisted/names/topfiles/4540.bugfix.  This should be really easy :-)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_5"></a>|@jonathanj set owner to @fijal|
|-|-|

1. The docstring for `searchFileFor` has a few issues:
   1. Docstrings should have the triple-quotes each on a line of their own.
   2. The grammar in docstring `searchFileFor` could probably be clearer about the fact that `file` should be in hosts(5) standard form.
   3. It looks like "give" should be "given".
2. `test_searchFileFor` needs a docstring.
3. Can you make sure there are 3 newlines before `FakeDNSDatagramProtocol` in `test_names.py`?

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_6"></a>|@jonathanj commented|
|-|-|

I forgot to mention: There should be tests for the cases where `searchFileFor` returns `None`.

Also, it looks like there are a bunch of other whitespace issues (mostly the number of newlines between classes and methods), refer to <http://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html>.

**Comment by twisted-trac** [comment]
|[<img alt="fijal's avatar" src="https://avatars.githubusercontent.com/u/85944?s=50" width="50" height="50">](https://github.com/fijal)<a name="note_7"></a>|@fijal removed owner|
|-|-|

Fixed reviewer notes (except whitespace a bit).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [29588]) Branching to 'close-hosts-4540'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [29589]) Apply out.2.diff

refs [#4540](https://github.com/twisted/twisted/issues/4540)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [29590]) add news fragment too

refs [#4540](https://github.com/twisted/twisted/issues/4540)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

I fixed some minor issues with the patch, whitespace, copyright dates, docstrings, and an assert method (r29590, r29591, r29592, r29593).  The rest seems fine to me and the [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/close-hosts-4540) are good, merging.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `closed`|
|-|-|

(In [29594]) Merge close-hosts-4540

Author: fijal, exarkun
Reviewer: mwh, jonathanj, exarkun
Fixes: [#4540](https://github.com/twisted/twisted/issues/4540)

Explicitly close the hosts file in twisted.names.hosts.searchFileFor, rather
than relying on reference counting to do it for us in a timely manner.  This
improves behavior on PyPy and other runtimes without reference counting.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
